src/gordon_gcp/plugins/publisher.py

```diff
@@ -13,44 +13,222 @@
+        self.http_client = http_client
+        self.publish_wait_timeout = config['publish_wait_timeout']
+        self.managed_zone = config['managed_zone']
+        self.dns_zone = config['dns_zone']
+        self.project = config['project']
+        self.default_ttl = self.config['default_ttl']
+        api_version = config.get('version', 'v1')
+        self.base_changes_url = BASE_CHANGES_ENDPOINT.format(
+            host=HOST, version=api_version, project=self.project,
+            managedZone=self.managed_zone)
+        self.resource_records_url = RESOURCE_RECORDS_ENDPOINT.format(
+            host=HOST, project=self.project, managedZone=self.managed_zone)
+        self._logger = logging.getLogger('')
+
+    # TODO: This will be eventually moved to GEventMessage
+    async def update_phase(self, event_msg, phase=None):
+        old_phase = event_msg.phase
+        event_msg.phase = phase or self.phase
+        msg = f'Updated phase from "{old_phase}" to "{event_msg.phase}".'
+        event_msg.append_to_history(msg, self.phase)
+
+    def _format_changes(self, action, resource_record):
+        """Return dict containing the changes to be made.
+
+        Args:
+            action (str): The action of the changes to publish;
+                can be one of ['additions', 'deletions'].
+            resource_record (dict): Dict containing record data.
+
+        Returns:
+            The changes (dict) to post to the Google API.
+        """
+        resource_record['kind'] = 'dns#resourceRecordSet'
+        resource_record['ttl'] = resource_record.get('ttl', self.default_ttl)
+        return {
+            'kind': 'dns#change',
+            action: [resource_record]
+        }
+
+    async def _publish_changes(self, changes):
+        """Publish changes to the Google DNS Changes API.
+
+        Args:
+            changes (dict): The DNS record changes to publish.
+
+        Returns:
+            change_id (str): Change ID of the Changes resource.
+
+        Raises:
+            GCPHTTPConflictError on HTTP 409.
+        """
+        try:
+            resp = await self.http_client.request(
+                'post', self.base_changes_url, json=changes)
+        except exceptions.GCPHTTPError as e:
+            error_msg = e.args[0]
+            # Extract status code from error message.
+            match = re.search('\d+', error_msg)
+            if match:
+                status_code = int(match.group(0))
+                if status_code == 409:
+                    raise exceptions.GCPHTTPConflictError(e)
+                elif status_code == 404:
+                    raise exceptions.GCPHTTPNotFoundError(e)
+            raise e
+
+        resp_dict = json.loads(resp)
+
+        # TODO: create another task to measure propagation time
+        return resp_dict['id']
+
+    async def _handle_additions_conflict(self, changes):
+        """Compensate for an addition Changes request that returned an HTTP 409.
```

Comment:
I would keep it in the present tense, i.e. `returns`. Also the longer description isn't really a sentence

Reply:
Oh yeah, and I think we were going to add a warning that method calls a URL that returns *ALL* rrsets in a zone, so it could potentially be huge and slow.

Reply:
I would also make it 'additions', since that's what Google calls it.